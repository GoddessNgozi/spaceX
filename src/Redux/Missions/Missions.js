import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const FETCH_MISSIONS = 'store/rockets/FETCH_MISSIONS';

const fetchMissions = createAsyncThunk(FETCH_MISSIONS, async () => {
  const response = await fetch('https://api.spacexdata.com/v3/missions');
  const missions = await response.json();
  return missions;
});

const missionSlice = createSlice({
  name: 'missions',
  initialState: [],
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchMissions.fulfilled, (state, action) => {
        let newState = state;
        newState = action.payload;
        return newState;
      })
      .addDefaultCase((state) => state);
  },
});

const missionsReducer = missionSlice.reducer;
export default missionsReducer;
export { fetchMissions };
