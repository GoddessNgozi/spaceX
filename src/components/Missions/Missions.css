table {
  width: calc(100% - 6rem);
  border: 1px solid rgba(128, 128, 128, 0.495);
  margin: 1rem 3rem;
}

th {
  text-align: left;
}

table,
th,
td {
  border: 1px solid rgba(128, 128, 128, 0.495);
  border-collapse: collapse;
  padding: 0.4rem;
}

tr:nth-of-type(2n + 2) {
  background-color: rgba(224, 220, 220, 0.418);
}
