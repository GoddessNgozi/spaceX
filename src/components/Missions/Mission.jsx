import Badge from "react-bootstrap/Badge";
import PropTypes from "prop-types";
import Button from "react-bootstrap/Button";
import { useDispatch } from "react-redux";
import { joinMission, leavingMission } from "../../Redux/Missions/Missions";

function Mission({ id, name, description, reserved }) {
  const dispatch = useDispatch();

  const join = () => {
    dispatch(joinMission(id));
  };

  const leave = () => {
    dispatch(leavingMission(id));
  };

  return (
    <>
      <tr key={id}>
        <td className="border">{name}</td>
        <td className="border" style={{maxWidth: '40rem'}}>{description}</td>
        <td className="border">
          <Badge bg={reserved ? "info" : "secondary"}>
            {reserved ? "ACTIVE MEMBER" : "NOT A MEMEBR"}
          </Badge>
        </td>
        <td 
          className="border"
          >
          <Button
            id={id}
            onClick={reserved ? leave : join}
            variant={reserved ? "outline-danger" : "outline-secondary"}
          >
            {reserved ? "Leave Misson" : "Join Mission"}
          </Button>
        </td>
      </tr>
    </>
  );
}

Mission.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  reserved: PropTypes.bool,
};

export default Mission;
