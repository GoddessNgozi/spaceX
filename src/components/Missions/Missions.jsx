import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Mission from './Mission';
import { fetchMissions } from '../../Redux/Missions/Missions';
import './Missions.css';

function Missions() {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);

  const missions = useSelector((state) => state.missions);
  return (
    <main>
      <table>
        <tbody>
          <tr>
            <th className="border">Mission</th>
            <th className="border">Description</th>
            <th className="border">Status</th>
            <th className="border">{' '}</th>
          </tr>
          {missions.map((mission) => (
            <Mission
              key={mission.mission_id}
              id={mission.mission_id}
              name={mission.mission_name}
              description={mission.description}
              reserved={mission.reserved}
            />
          ))}
        </tbody>
      </table>
    </main>
  );
}

export default Missions;
