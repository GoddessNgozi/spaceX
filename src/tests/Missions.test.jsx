import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import Missions from '../components/Missions/Missions';
import { fetchMissions, joinMission, leavingMission } from '../redux/Missions/Missions';

describe('Missions renders correctly', () => {
  test('Renders missions Correctly', () => {
    const missions = render(
      <Provider store={store}>
        <Router>
          <Missions />
        </Router>
      </Provider>,
    );
    expect(missions).toMatchSnapshot();
  });
});

describe('Missions reducers', () => {
  it('should fetch missions', async () => {
    await store.dispatch(fetchMissions());
    expect(store.getState().missions.length).toEqual(10);
  });

  it('should reserve a missions', () => {
    store.dispatch(joinMission('9D1B7E0'));
    expect(store.getState().missions[0].reserved).toBe(true);
  });

  it('should cancel a reservation', () => {
    store.dispatch(leavingMission('9D1B7E0'));
    expect(store.getState().missions[0].reserved).toEqual(false);
  });
});
