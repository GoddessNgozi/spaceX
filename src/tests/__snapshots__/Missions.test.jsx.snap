// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions renders correctly Renders missions Correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main>
        <table>
          <tbody>
            <tr>
              <th
                class="border"
              >
                Mission
              </th>
              <th
                class="border"
              >
                Description
              </th>
              <th
                class="border"
              >
                Status
              </th>
              <th
                class="border"
              >
                 
              </th>
            </tr>
          </tbody>
        </table>
      </main>
    </div>
  </body>,
  "container": <div>
    <main>
      <table>
        <tbody>
          <tr>
            <th
              class="border"
            >
              Mission
            </th>
            <th
              class="border"
            >
              Description
            </th>
            <th
              class="border"
            >
              Status
            </th>
            <th
              class="border"
            >
               
            </th>
          </tr>
        </tbody>
      </table>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
